runtimebp/cpu.go

```diff
@@ -12,75 +12,120 @@ import (
 // NumCPU returns the number of CPUs assigned to this running container.
 //
 // This is the container aware version of runtime.NumCPU.
-// It reads from the cgroup cpu.cfs_quota_us and cpu.cfs_period_us values
-// to determine the hard CPU limit of the container.
+// It reads from the cgroup v2 cpu.max values to determine the hard CPU limit of
+// the container.
 //
+// If the current process is not running with cgroup v2,
+// it falls back to read from the cgroup v1 cpu.cfs_quota_us and
+// cpu.cfs_period_us values.
 // If the current process is not running inside a container,
 // or if there's no limit set in cgroup,
 // it will fallback to runtime.NumCPU() instead.
 func NumCPU() float64 {
+	// Big enough buffer to read the numbers in the files wholly into memory.
+	buf := make([]byte, 1024)
+
+	n, err := numCPUCgroupsV2(buf)
+	if err == nil {
+		return n
+	}
+
+	// fallback to cgroups v1
+	fmt.Fprintf(
+		os.Stderr,
+		"NumCPU: Failed to read cgroup v2, fallback to cgroup v1: %v\n",
+		err,
+	)
+	n, err = numCPUCgroupsV1(buf)
+	if err == nil {
+		return n
+	}
+
+	// Fallback to the standard runtime package value.
+	fmt.Fprintf(
+		os.Stderr,
+		"NumCPU: Failed to read cgroup v1, fallback to NumCPU on the physical machine: %v\n",
+		err,
+	)
+	return float64(runtime.NumCPU())
+}
+
+func numCPUCgroupsV2(buf []byte) (float64, error) {
 	const (
-		quotaPath  = "/sys/fs/cgroup/cpu/cpu.cfs_quota_us"
-		periodPath = "/sys/fs/cgroup/cpu/cpu.cfs_period_us"
+		maxPath = "/sys/fs/cgroup/cpu.max"
 	)
 
-	// Default to the standard runtime package value.
-	defaultCPUs := float64(runtime.NumCPU())
+	values, err := readNumbersFromFile(maxPath, buf)
+	if err != nil || len(values) != 2 {
+		return 0, fmt.Errorf(
+			"runtimebp: failed to read max file, falling back to use runtime.NumCPU(): %w, %v",
```

Comment:
Only include fallback messaging when you actually execute the fallback logic.  I'm pretty sure this is going to be printed multiple times because it's returned from the `num...V1/V2` functions and in the driver function.

Reply:
right, I forgot to remove the fallback message when converting this from print to stderr to constructing error message. fixed.